Sources/SourceKitLSP/TestDiscovery.swift

```diff
@@ -462,24 +485,131 @@ extension TestItem {
+  /// - `MyTests` (FileB.swift:1)
+  ///   - `twoIsThree`
+  ///
+  /// A node's parent is identified by the node's ID with the last component dropped.
+  func mergingTestsInExtensions() -> [TestItem] {
+    var itemDict: [String: AnnotatedTestItem] = [:]
+    for item in self {
+      let id = item.testItem.id
+      if var rootItem = itemDict[id] {
+        // If we've encountered an extension first, and this is the
+        // type declaration, then use the type declaration TestItem
+        // as the root item.
+        if rootItem.isExtension && !item.isExtension {
+          var newItem = item
+          newItem.testItem.children += rootItem.testItem.children
+          rootItem = newItem
+        } else {
+          rootItem.testItem.children += item.testItem.children
+        }
+
+        itemDict[id] = rootItem
+      } else {
+        itemDict[id] = item
+      }
+    }
+
+    if itemDict.isEmpty {
+      return []
+    }
+
+    var mergedIds = Set<String>()
+    for item in self {
+      let id = item.testItem.id
+      let parentID = id.components(separatedBy: "/").dropLast().joined(separator: "/")
+      // If the parent exists, add the current item to its children and remove it from the root
+      if var parent = itemDict[parentID] {
+        parent.testItem.children.append(item.testItem)
+        mergedIds.insert(parent.testItem.id)
+        itemDict[parent.testItem.id] = parent
+        itemDict[id] = nil
+      }
+    }
+
+    // Filter out the items that have been merged into their parents, sorting the tests by location.
+    // TestItems not in extensions should be priotitized first.
+    let sortedItems = itemDict.values
+      .compactMap { $0 }
+      .sorted { ($0.isExtension != $1.isExtension) ? !$0.isExtension : ($0.testItem.location < $1.testItem.location) }
+
+    return sortedItems.map {
+      guard !$0.testItem.children.isEmpty, mergedIds.contains($0.testItem.id) else {
+        return $0.testItem
+      }
+      var newItem = $0.testItem
+      newItem.children = newItem.children
+        .map { AnnotatedTestItem(testItem: $0, isExtension: false) }
+        .mergingTestsInExtensions()
+      return newItem
+    }
+  }
+}
+
 extension SwiftLanguageService {
-  public func syntacticDocumentTests(for uri: DocumentURI, in workspace: Workspace) async throws -> [TestItem]? {
+  public func syntacticDocumentTests(for uri: DocumentURI, in workspace: Workspace) async throws -> [AnnotatedTestItem]?
+  {
     let snapshot = try documentManager.latestSnapshot(uri)
     let semanticSymbols = workspace.index(checkedFor: .deletedFiles)?.symbols(inFilePath: snapshot.uri.pseudoPath)
     let xctestSymbols = await SyntacticSwiftXCTestScanner.findTestSymbols(
       in: snapshot,
       syntaxTreeManager: syntaxTreeManager
-    ).compactMap { $0.filterUsing(semanticSymbols: semanticSymbols) }
+    )
+    .compactMap { $0.filterUsing(semanticSymbols: semanticSymbols) }
+    .map { AnnotatedTestItem(testItem: $0, isExtension: false) }
```

Comment:
Similar here, XCTests can also have extensions. I would be fine with a follow-up PR for extension merging of XCTest tests, if you would prefer that.